kmm/kmm-post-data/src/commonMain/kotlin/com/gchristov/newsfeed/kmmpostdata/usecase/DecoratePostUseCase.kt

```diff
@@ -10,19 +10,49 @@ import kotlinx.datetime.Instant
 
 class DecoratePostUseCase(
     private val postRepository: PostRepository,
-    private val dispatcher: CoroutineDispatcher) {
+    private val dispatcher: CoroutineDispatcher
+) {
 
-    suspend fun decoratedPost(
-        postId: String
-    ): DecoratedPost = fetchDecoratedPost(postId)
 
-    suspend fun cachedPost(postId: String): DecoratedPost? =
-        postRepository.cachedPost(postId)?.let { post ->
-            clearCache(post.id)
-            decoratePost(post)
+    /**
+     * Obtain a new post from API/repository, decorate and
+     * cache it.
+     * Optionally provide a callback that eagerly retrieves/checks
+     * the posts cache
+     *
+     * @param postId the ID of the post to obtain
+     * @param onCache callback that pulls posts from cache if existing
+     */
+    suspend fun getPost(
+        postId: String,
+        // the whole callback is nullable/optional
+        onCache: ((DecoratedPost) -> Unit)?
+    ): DecoratedPost {
+
+        // We wouldn't even run this if the caller doesn't care about cache
+        onCache?.let { cacheCallback ->
+            cachedPost(postId)?.let { post ->
+
+                // this is normally to update the UI from cache while
+                // a new one loads
+                cacheCallback(post)
+            }
         }
 
-    suspend fun clearCache(postId: String) = postRepository.clearCache(postId)
+        return fetchDecoratedPost(postId)
+    }
+
+    /**
+     * This method is private to force callers to either handle cache or not explicitly
+     * via callback
+     */
+    private suspend fun cachedPost(postId: String): DecoratedPost? =
+        postRepository.run {
+            cachedPost(postId)?.let { post ->
+                clearCache(post.id)
```

Comment:
If we delete the cache after we use it here and then there’s a network failure afterwards for example we won’t have a cache to show the next time the user opens the post. 

I’d put the clearCache after we have downloaded the post because we know for sure it’s available at that time.

Reply:
That's great insight - I love this pattern of responsiveness!

Reply:
This should've been done so feel free to resolve